Sneak peek: API documentation generated for all your code (#3815)
sourcegraph/about: blogposts/2021/api-documentation-generated-for-all-your-code.md

</figure>

### A deep understanding of code

We're paying attention to the small details, and really tailoring API docs to each language.

At Sourcegraph, we use primarily Go and TypeScript—so we're starting by building an experience that Go developers will love, then we'll expand to more languages and build an experience that is really tailored to, and makes sense in the context of, the individual language.

One way we've enabled this is by allowing LSIF language indexers to emit data in the way they see fit. Sourcegraph merely renders it, indexes it, and adds a nice navigation experience.

<figure>
  <img src="https://sourcegraphstatic.com/blog/api-docs/url-hash-displays.gif" alt="Browser URL bar showing a URL and hash designed specifically for the Go language">
  <figcaption>The Go LSIF indexer chooses to display URL hashes in a way that makes sense specifically for Go, with a `Receiver.Method` format</figcaption>
</figure>

<figure>
  <img height="400px" src="https://sourcegraphstatic.com/blog/api-docs/symbol-types.gif" alt="Hovering over symbols and showing Sourcegraph's deep semantic understanding of type information">

>   <figcaption>Hovering over symbols shows Sourcegraph's deep semantic understanding of type information, which you'll soon be able to search and filter based on.</figcaption>

Author: @slimsag The caption ends quite abruptly, is there a way you can reword it? It might just be the ending with "on" which always sounds a bit awkward 😅 

I've also edited the other captions so they're consistent with no periods at the end.

Reviewer: ```suggestion
  <figcaption>Soon it will be possible to search and filter using Sourcegraph's deep semantic understanding of type information</figcaption>
```

---
title: "Sneak peek: API documentation generated for all your code"
description: "Here's a sneak peek of our latest experiment to bring API documentation to all your code, generated by LSIF code intelligence data."
author: Stephen Gutekanst
authorUrl: 
publishDate: 2021-07-21T10:00-07:00
tags: [blog, code, search, software, engineering, testing]
slug: api-documentation-for-all-your-code

> heroImage: https://sourcegraphstatic.com/blog/api-docs/api-docs.png

Reviewer: I put together something we can use [here](https://user-images.githubusercontent.com/3173176/126405059-ce489151-c7eb-4dbf-87f1-1f73a0000141.png) just now:

![image](https://user-images.githubusercontent.com/3173176/126405059-ce489151-c7eb-4dbf-87f1-1f73a0000141.png)

socialImage: https://sourcegraphstatic.com/blog/api-docs/api-docs.png 
published: true
---

At Sourcegraph, we have a small team working hard on a new experimental feature: API docs.

> Using [LSIF code intelligence data](https://lsif.dev), we can analyze your code and generate API documentation for all of it. We've started with Go code and intend to expand to more languages as we develop this feature. 

Author: I think it's worth including a very brief description of Sourcegraph here. I try to imagine someone landing on this page from HN with no idea who we are or what we do, and think of what context would be helpful before diving in. 

I can take a stab, but I'd prefer an intro that connects the dots from what Sourcegraph does/offers to API docs ("we do X at Sourcegraph which gives us insight into Y, therefore we can generate API docs for you based on Y" – this is bad but I think you get the idea??)

> Using [LSIF code intelligence data](https://lsif.dev), we can analyze your code and generate API documentation for all of it. We've started with Go code and intend to expand to more languages as we develop this feature. 

Reviewer: How about something like this maybe?

```suggestion
At Sourcegraph, in order to build awesome code search we've built a global graph of code that has a deep, semantic understanding of the underlying code - which we represent as LSIF code intelligence data. Using Sourcegraph's [LSIF code intelligence data](https://lsif.dev), we can analyze your code and generate API documentation for all of it. We've started with Go code and intend to expand to more languages as we develop this feature. 
```

> ![API Docs](https://sourcegraphstatic.com/blog/api-docs/api-docs.png)

Author: @slimsag Sorry I forgot to mention this before, could you please add captions to the screengrabs? If someone is skimming the post we have a better chance of drawing them in if they can see at a glance what an image or gif is about. You can see some examples here: https://about.sourcegraph.com/blog/optimizing-a-code-intel-commit-graph-part-2/

Also, the images aren't consistent sizes (the feedback widget one is comically large 😆) and I think the only way to apply sizing to them is with HTML. This is not super familiar for me but if you know what you're doing please go ahead and make them consistent!

Reviewer: Done, I updated all locations! As for the HTML stuff, in case you're curious to learn - the trick is just to write the image like this instead:

```html
<img height="50px" src="https://sourcegraphstatic.com/blog/api-docs/api-docs.png"></img>
```

(and if you just set `height`, the image width will be automatically adjusted)

## Try it (experimental)

Today, it's still in its very early stages—and only a few months into development. But you can try it out on Go code today:

- [Explore encoding/json code](https://sourcegraph.com/github.com/golang/go/-/docs/encoding/json)
- [Explore Go net/http code](https://sourcegraph.com/github.com/golang/go/-/docs/net/http)
- [Explore Sourcegraph backend code](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/docs/cmd/frontend/backend)

Want to try it out on your own Go repositories? Tweet [@sourcegraph](https://twitter.com/sourcegraph) with your GitHub repository name and we'll set it up for you, or [follow our documentation]([https://docs.sourcegraph.com/code_intelligence/apidocs](https://docs.sourcegraph.com/code_intelligence/apidocs)).

## Improving rapidly

Here's a little highlight of the changes we've made recently.

### On by default, for everyone

The most important announcement here is that API docs is now an on-by-default experiment for everyone! You can begin trying it out on Sourcegraph.com as well as your own Sourcegraph instance by clicking on the API docs button on any repository page:

> ![API Docs button](https://sourcegraphstatic.com/blog/api-docs/api-docs-button.png)

Author: Caption

Once you click it, you'll be greeted with information about how to get started in a brand new welcome prompt we’ve added:

> ![Welcome prompt](https://sourcegraphstatic.com/blog/api-docs/welcome-prompt.png)

Author: Caption

### Better sub-page navigation

Within a Go package, there are many sub-packages. We've improved the way we render documentation subpages so you can skip over useless directories and get right to the content (but we're still hard at work improving navigation in API docs!):

> ![Sub-page navigation](https://sourcegraphstatic.com/blog/api-docs/subpage-navigation.gif)

Author: Caption

### Understands the types of symbols, and uses iconography to communicate

Due to API docs being built on code intelligence data, it understands exactly what symbols are and has detailed information about them. Whether symbols are deprecated, whether they're a function, variable, interface, etc. and the UI now communicates this information using icons:

> ![Symbols information](https://sourcegraphstatic.com/blog/api-docs/symbols.gif)

Author: Caption

### A deep understanding of code

> We're paying attention to the small details, and really tailoring API docs to each language. We want an experience that makes sense for Go first, then we'll expand to more languages and build an experience that makes sense for another language.

Author: I asked this in the G Doc, are we choosing Go first because we use Go to develop Sourcegraph? Someone unfamiliar with our stack might wonder why Go.

Reviewer: ```suggestion
We're paying attention to the small details, and really tailoring API docs to each language. We want an experience that makes sense for Go first, because here at Sourcegraph we use primarily Go and TypeScript and want to ensure we build an experience Go developers love, then we'll expand to more languages and build an experience that makes sense in the context of other languages - a tailored experience for each.
```

One way we've enabled this is by allowing LSIF language indexers to emit data in the way they see fit. Sourcegraph merely renders it, indexes it, and adds a nice navigation experience. For example, the Go LSIF indexer even chooses to display URL hashes in a way that makes sense specifically for Go, with a `Receiver.Method` format:

![URL hash display for Go](https://sourcegraphstatic.com/blog/api-docs/url-hash-displays.gif)

> And it understands symbol types, too—just displaying them for now, but enabling you to search and view e.g. "just variables", "just functions", etc. in the near future:

Author: How is this different to what you describe in `Understands the types of symbols, and uses iconography to communicate` above (60-64)?

Reviewer: Yeah, you're right, it's pretty duplicative actually. The point here is more about "We'll use this to bring you some cooler features in the future" - how about this:

```suggestion
In the near future, we'll enable you to search and filter based on symbol types as well. "just show me variables", "just private and deprecated functions", etc. based on the deep understanding of the code we've built:
```

> ![URL hash display for Go](https://sourcegraphstatic.com/blog/api-docs/url-hash-displays.gif)

Author: Caption

> ![Display symbol types](https://sourcegraphstatic.com/blog/api-docs/symbol-types.gif)

Author: Caption

### Feedback widget

We know there will be a ton of feedback as things are very early in development, and so we've made it as easy as possible to give feedback from any API docs page:

> ![Feedback widget](https://sourcegraphstatic.com/blog/api-docs/feedback-widget.png)

Author: Caption

> You can also directly email feedback to the lead engineer [stephen@sourcegraph.com](mailto:stephen@sourcegraph.com) (that's me) or the entire team via Twitter [@sourcegraph](https://twitter.com/sourcegraph). We really want to hear what you think!

Author: @slimsag Do you want to add an author bio for yourself (example [here](https://about.sourcegraph.com/blog/how-not-to-break-a-search-engine-unglamorous-engineering/))? It's optional so up to you!

Reviewer: Done!
